### ISSUE TITLE
feat: Auto-compute hamiltonian_dm_func via finite-diff or autodiff

### ISSUE THREAD

**Issue by derrring** [derrring/MFG_PDE#667]
## Summary

When users provide `hamiltonian_func` without `hamiltonian_dm_func`, automatically compute the derivative using finite differences or autodiff (JAX/PyTorch if available).

## Current Behavior

- If `hamiltonian_dm_func` is not provided, a **zero placeholder** is used with a warning
- This is incorrect for Hamiltonians with density dependence (e.g., congestion terms like `-λ·log(m)`)

## Proposed Behavior

```python
# Option 1: Explicit auto-computation
components = MFGComponents(
    hamiltonian_func=my_H,
    hamiltonian_dm_func="auto",  # Finite-diff or autodiff
)

# Option 2: Smart default (when hamiltonian_dm_func is None)
# Detect if JAX available → use autodiff
# Otherwise → use finite-diff
# Emit info log about which method is used
```

## Mathematical Background

For **local Hamiltonians** (current API):
```
H(x, p, m) where m = m(x) is scalar (local density)
dH/dm = ∂H/∂m  — simple partial derivative
```

**Finite-diff implementation:**
```python
def auto_dH_dm(H, x_idx, x_pos, m, derivs, t_idx, t, problem, eps=1e-7):
    H_plus = H(x_idx, x_pos, m + eps, derivs, t_idx, t, problem)
    H_minus = H(x_idx, x_pos, m - eps, derivs, t_idx, t, problem)
    return (H_plus - H_minus) / (2 * eps)
```

**Autodiff implementation (JAX):**
```python
import jax

def create_autodiff_dH_dm(H):
    def dH_dm(x_idx, x_pos, m, derivs, t_idx, t, problem):
        def H_of_m(m_val):
            return H(x_idx, x_pos, m_val, derivs, t_idx, t, problem)
        return jax.grad(H_of_m)(m)
    return dH_dm
```

## Accuracy Comparison

| Method | Error vs Analytic | Notes |
|--------|------------------|-------|
| Finite-diff (ε=1e-7) | ~1e-10 | Central difference, O(ε²) |
| JAX autodiff | ~1e-8 | Machine precision for float32 |

## Implementation Notes

1. Add `hamiltonian_dm_mode` parameter to `MFGComponents` or `_validate_hamiltonian_components`
2. Finite-diff as default fallback (no dependencies)
3. Prefer JAX/PyTorch autodiff if available and user opts in
4. Log which method is used

## Out of Scope

- **Functional derivatives** for non-local Hamiltonians (H depends on entire m field)
- Would require API change: `m_at_x: float` → `m_field: NDArray`
- Track in separate issue if needed

## Acceptance Criteria

- [ ] Finite-diff auto-computation works correctly
- [ ] Warning emitted when using auto-computation (user should verify)
- [ ] Optional JAX autodiff backend
- [ ] Tests comparing auto vs analytic for Towel-on-Beach
